frontend/packages/react-app/src/App.tsx

```diff
@@ -1,31 +1,88 @@
-import React, { Component } from 'react';
-import { BrowserRouter as Router, Route } from 'react-router-dom';
-import TopPage from "./components/pages/TopPage";
+import React, {
+  createContext,
+  useCallback,
+  useEffect,
+  useReducer,
+  useContext,
+} from "react";
+import { BrowserRouter as Router, Route } from "react-router-dom";
 import CreateTokenPage from "./components/pages/CreateTokenPage";
 import ExplorePage from "./components/pages/ExplorePage";
 import DashboardPage from "./components/pages/DashboardPage";
 import TokenDetailPage from "./components/pages/TokenDetailPage";
 import WithdrawAudiusPage from "./components/pages/WithdrawAudiusPage";
-import AudiusDisributionPage from "./components/pages/AudiusDisributionPage"
-
-class App extends Component {
-  render() {
-    return (
-      <div>
-        <Router>
-          <div>
-            <Route exact path='/' component={TopPage}/>
-            <Route path='/create' component={CreateTokenPage}/>
-            <Route path='/explore' component={ExplorePage}/>
-            <Route path='/dashboard' component={DashboardPage}/>
-            <Route path='/token/:address' component={TokenDetailPage}/>
-            <Route path='/audius-token' component={WithdrawAudiusPage}/>
-            <Route path='/audius/:address' component={AudiusDisributionPage}/>
-          </div>
-        </Router>
-      </div>
-    );
+import AudiusDisributionPage from "./components/pages/AudiusDisributionPage";
+
+import { Web3Provider } from "@ethersproject/providers";
+import { web3Modal } from "./utils/web3Modal";
+
+type State = {
+  count: number;
+  provider: Web3Provider | undefined;
+};
+type Action = { type: "update_provider"; payload: { provider: Web3Provider } };
+
+const reducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case "update_provider":
+      return { ...state, provider: action.payload.provider };
+    default:
+      return state;
+  }
+};
+
+export const AppContext = createContext(
+  {} as {
+    state: State;
+    dispatch: React.Dispatch<Action>;
   }
+);
+const initialState = {
+  count: 0,
+  provider: undefined,
+};
+
+const UpdateWeb3Context = createContext({});
+export function useUpdateWeb3() {
+  return useContext(UpdateWeb3Context);
 }
```

Comment:
- Could you extract reducer and context as a single files like `src/reducers/web3.ts` and `src/contexts/web3.ts`respectively? 
Hooks types is provided like `React.Dispatch`.
- You can define type used in context with generics `<>` like following: 
```javascript 
// Reducer
type AppAction = { type: "update_provider"; payload: { provider: Web3Provider } };

type AppState = {
  provider: Web3Provider | undefined,
};

const initialState: AppState = {
  provider: undefined,
};

const reducer = (state: AppState, action: AppAction) => {
  switch (action.type) {
    case "update_provider":
      return { ...state, provider: action.payload.provider };
    default:
      return state;
  }
};

// Context
interface AppContext {
  state: AppState,
  dispatch: React.Dispatch<AppAction>
}

export const AppContext = createContext<AppContext>({
  state: initialState,
  dispatch: () => initialState,
});

const UpdateWeb3Context = createContext({});
export function useUpdateWeb3() {
  return useContext(UpdateWeb3Context);
}
```

Reply:
OK